lib/compiler/runtime.ts

```diff
@@ -0,0 +1,102 @@
+import { readFileSync } from 'fs';
+import { dirname } from 'path';
+import { transpileModule, TranspileOptions } from 'typescript';
+import { RunningScriptOptions } from 'vm';
+import { Container } from './container';
+import { Compiler } from './index';
+
+export class Runtime {
+  private readonly transpileOptions: TranspileOptions;
+
+  private readonly containerRequireCache: { [name: string]: any } = {};
+
+  constructor(compiler: Compiler) {
+    this.transpileOptions = {
+      compilerOptions: compiler.project.compilerOptions.get()
+    };
+  }
+
+  /**
+   * @internal
+   */
+  public createContainer<T>(container: Container<T>): T {
+    const output = transpileModule(container.code, this.transpileOptions);
+    const code = output.outputText;
+
+    const evaluatedExports = this.evaluateModule(
+      code,
+      `<${container.implName}.ts>`,
+      {
+        lineOffset: 2
+      }
+    );
+
+    const constructor: { new (): T } = evaluatedExports[container.implName];
+    return new constructor();
+  }
+
+  public require(id: string, from: string): any {
+    const resolvedId = require.resolve(id, {
+      paths: [dirname(from)]
+    });
+    if (resolvedId.endsWith('.js')) {
+      return require(resolvedId);
+    }
+
+    if (this.containerRequireCache[resolvedId] !== undefined) {
+      return this.containerRequireCache[resolvedId];
+    }
+
+    const code = transpileModule(
+      readFileSync(resolvedId).toString(),
+      this.transpileOptions
+    ).outputText;
+    const exports = this.evaluateModule(code, resolvedId);
+
+    this.containerRequireCache[resolvedId] = exports;
+    return this.containerRequireCache[resolvedId];
+  }
+
+  private evaluateModule(
```

Comment:
To create containers directly in the process of the compiler the generated code need to be evaluated. The main purpose for now is to use this in unit tests and its internal api and will not be bundled in a browser version.
Otherwise it would be necessary to either write the code to disk and then start another process to run it or to move the `hacky` evaluation part into the test code. The second way was integrated in the first compiler iteration which will be overwritten with this PR and has had a lot more flaws since it could not access the internals of the container.

So in summary: yes its a hack, but its a controlled hack for special runtime cases like tests.